backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/search/projection/impl/DistanceFieldSearchProjectionImpl.java

```diff
@@ -0,0 +1,98 @@
+/*
+ * Hibernate Search, full-text search for your domain model
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package org.hibernate.search.backend.elasticsearch.search.projection.impl;
+
+import java.util.Optional;
+import java.util.regex.Pattern;
+
+import org.hibernate.search.backend.elasticsearch.gson.impl.JsonAccessor;
+import org.hibernate.search.backend.elasticsearch.gson.impl.JsonObjectAccessor;
+import org.hibernate.search.engine.search.query.spi.ProjectionHitCollector;
+import org.hibernate.search.engine.spatial.DistanceUnit;
+import org.hibernate.search.engine.spatial.GeoPoint;
+
+import com.google.gson.JsonObject;
+
+public class DistanceFieldSearchProjectionImpl implements ElasticsearchSearchProjection<Double> {
+
+	private static final JsonObjectAccessor SCRIPT_FIELDS_ACCESSOR = JsonAccessor.root().property( "script_fields" ).asObject();
+	private static final JsonObjectAccessor FIELDS_ACCESSOR = JsonAccessor.root().property( "fields" ).asObject();
+
+	private static final Pattern NON_DIGITS_PATTERN = Pattern.compile( "\\D" );
+
+	private final String absoluteFieldPath;
+
+	private final GeoPoint center;
+
+	private final DistanceUnit unit;
+
+	private final String scriptFieldName;
+
+	DistanceFieldSearchProjectionImpl(String absoluteFieldPath, GeoPoint center, DistanceUnit unit) {
+		this.absoluteFieldPath = absoluteFieldPath;
+		this.center = center;
+		this.unit = unit;
+		this.scriptFieldName = createScriptFieldName( absoluteFieldPath, center, unit );
+	}
+
+	@Override
+	public void contributeRequest(JsonObject requestBody) {
+		SCRIPT_FIELDS_ACCESSOR
+				.property( scriptFieldName ).asObject()
+				.property( "script" ).asObject()
+				.set( requestBody, createScript( absoluteFieldPath, center ) );
+	}
+
+	@Override
+	public void extract(ProjectionHitCollector collector, JsonObject responseBody, JsonObject hit) {
+		Optional<Double> distance = FIELDS_ACCESSOR.property( scriptFieldName ).asArray()
+				.element( 0 )
+				.asDouble().get( hit );
+
+		collector.collectProjection( distance.isPresent() && Double.isFinite( distance.get() ) ?
+				unit.metersToUnit( distance.get() ) : null );
+	}
+
+	@Override
+	public String toString() {
+		StringBuilder sb = new StringBuilder( getClass().getSimpleName() )
+				.append( "[" )
+				.append( "absoluteFieldPath=" ).append( absoluteFieldPath )
+				.append( ", center=" ).append( center )
+				.append( ", unit=" ).append( unit )
+				.append( "]" );
+		return sb.toString();
+	}
+
+	private static String createScriptFieldName(String absoluteFieldPath, GeoPoint center, DistanceUnit unit) {
+		StringBuilder sb = new StringBuilder();
+		sb.append( "distance_" )
+				.append( absoluteFieldPath )
+				.append( "_" )
+				.append( NON_DIGITS_PATTERN.matcher( Double.toString( center.getLatitude() ) ).replaceAll( "_" ) )
```

Comment:
Could you please add tests for the case I mentioned above, and also for cases where the doubles are really precise, with lots of digits, so that we're sure Elasticsearch won't throw some exception due to the field name being too long?

Reply:
I did some tests, and it seems I was wrong, as long as we only use doubles that make sense, there shouldn't be any problem. This is mainly because the plus sign is not displayed, so we can differentiate between positive and negative values based on the presence or absence of a "_" prefix.

The only collision I was able to find requires invalid coordinates: `(1.0, -44000000.0)` and `(0.0001, 4.7)` both translate to `1_0__4_4_7`.

If you want to experiment as well, here's the shell script I used:
```
JAVA9_HOME/bin/jshell <<EOF
Pattern nonDigit = Pattern.compile("\\\\D");
for ( double val : Arrays.asList( 1.0, -1.0, 1.0, -44000000.0, 0.0001, 4.7 ) ) {
    System.out.println("" + val + " => '" + nonDigit.matcher(Double.toString(val)).replaceAll("_") + "'");
}
EOF
```

Output:

```
1.0 => '1_0'
-1.0 => '_1_0'
1.0 => '1_0'
-4.4E7 => '_4_4_7'
1.0E-4 => '1_0__4'
4.7 => '4_7'
```

I still think we'd need a test to check that Elasticsearch is fine with very long field names, though.

Reply:
OK, to be honest, it wasn't designed to be entirely free of conflicts but be a reasonable way to differentiate the fields.

I added a test with a long field name.